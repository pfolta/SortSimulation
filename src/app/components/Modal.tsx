import React, { ReactNode } from "react";
import { X } from "react-feather";
import { useIntl } from "react-intl";
import ReactModal, { Props as ReactModalProps } from "react-modal";
import styled from "styled-components";

import { BackgroundFadeIn, BackgroundFadeOut, SlideIn, SlideOut } from "@/app/animations";

interface ReactModalAdapterProps extends ReactModalProps {
    children?: ReactNode;
    className?: string;
}

const ReactModalAdapter = ({ className, ...delegated }: ReactModalAdapterProps) => (
    <ReactModal portalClassName={className} className={`${className}__Content`} overlayClassName={`${className}__Overlay`} {...delegated} />
);

const ANIMATION_DURATION = 250;

const StyledReactModal = styled(ReactModalAdapter).attrs({
    closeTimeoutMS: ANIMATION_DURATION
})`
    &__Overlay {
        display: grid;
        place-items: end center;

        position: fixed;
        top: 0;
        left: 0;
        width: 100vw;
        height: 100vh;
        z-index: 2;

        @media (min-width: ${({ theme }) => theme.modal.maxWidth}) {
            place-items: center;
        }

        &.ReactModal__Overlay--after-open {
            animation: ${({ theme }) => BackgroundFadeIn(theme.modal.background)} ${ANIMATION_DURATION}ms ease-out forwards;
        }

        &.ReactModal__Overlay--before-close {
            animation: ${({ theme }) => BackgroundFadeOut(theme.modal.background)} ${ANIMATION_DURATION}ms ease-in forwards;
        }
    }

    &__Content {
        display: grid;
        position: relative;

        background: ${({ theme }) => theme.body.background};
        box-shadow: 0 0.5rem 2rem 0 rgba(0, 0, 0, 0.75);
        border-radius: 0.5rem 0.5rem 0 0;

        width: min(${({ theme }) => theme.modal.maxWidth}, 100vw);
        max-height: calc(100vh - ${({ theme }) => theme.modal.marginTop});

        outline: none;
        overflow: hidden;

        @media (min-width: ${({ theme }) => theme.modal.maxWidth}) {
            border-radius: 0.5rem;
        }

        & > * {
            padding: ${({ theme }) => theme.modal.padding};
            overflow: auto;
            max-height: calc(100vh - ${({ theme }) => theme.modal.marginTop} - 2 * ${({ theme }) => theme.modal.padding});
        }

        &.ReactModal__Content--after-open {
            animation: ${SlideIn} ${ANIMATION_DURATION}ms ease-out forwards;
        }

        &.ReactModal__Content--before-close {
            animation: ${SlideOut} ${ANIMATION_DURATION}ms ease-in forwards;
        }
    }
`;

const CloseButton = styled.button`
    all: unset;

    display: grid;
    padding: 0.25rem;

    position: absolute;
    top: 1rem;
    right: 1rem;

    background-color: ${({ theme }) => theme.colors.gray5};
    color: ${({ theme }) => theme.colors.gray3};

    border-radius: 0.25rem;

    :focus {
        box-shadow: ${({ theme }) => theme.focus.boxShadow};
    }

    :hover {
        color: ${({ theme }) => theme.colors.gray1};
    }

    :active {
        background-color: ${({ theme }) => theme.colors.gray4};
    }
`;

interface ModalProps {
    children?: ReactNode;
    isOpen: boolean;
    onClose: () => void;
}

const Modal = ({ children, onClose, isOpen }: ModalProps): JSX.Element => {
    const intl = useIntl();

    const disableBodyScroll = () => (document.body.style.overflow = "hidden");
    const enableBodyScroll = () => (document.body.style.overflow = "unset");

    return (
        <StyledReactModal isOpen={isOpen} onRequestClose={onClose} onAfterOpen={disableBodyScroll} onAfterClose={enableBodyScroll}>
            <CloseButton title={intl.formatMessage({ id: "close" })} onClick={onClose}>
                <X size="1.25rem" />
            </CloseButton>
            {children}
        </StyledReactModal>
    );
};

export default Modal;
