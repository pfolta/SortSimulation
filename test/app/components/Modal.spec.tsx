import React from "react";

import renderWithProviders from "../../utils/renderWithProviders";

import Modal from "@/app/components/Modal";

describe("Modal", () => {
    it("renders correctly", () => {
        const component = renderWithProviders(<Modal onClose={jest.fn} isOpen={true} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders the modal content if the modal is opened", () => {
        const content = <section>This is the modal content.</section>;

        const component = renderWithProviders(
            <Modal onClose={jest.fn} isOpen={true}>
                {content}
            </Modal>
        );

        expect(component.toJSON()).toMatchSnapshot();

        const modalContent = component.root.findByType("section");
        expect(modalContent).toBeDefined();
        expect(modalContent.children[0]).toBe("This is the modal content.");
    });

    it("does not render the modal content if the modal is closed", () => {
        const content = <section>This is the modal content.</section>;

        const component = renderWithProviders(
            <Modal onClose={jest.fn} isOpen={false}>
                {content}
            </Modal>
        );

        expect(component.toJSON()).toMatchSnapshot();
        expect(() => component.root.findByType("section")).toThrow(Error);
    });

    it("clicking the 'X' button triggers the `onClose` function", () => {
        const onCloseMock = jest.fn();

        const component = renderWithProviders(<Modal onClose={onCloseMock} isOpen={true} />);

        const closeButton = component.root.findByType("button");
        expect(closeButton).toBeDefined();

        closeButton.props.onClick();
        expect(onCloseMock).toHaveBeenCalledTimes(1);
    });
});
