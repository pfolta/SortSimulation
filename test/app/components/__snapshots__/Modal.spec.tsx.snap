// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal does not render the modal content if the modal is closed 1`] = `
@media (min-width:45rem) {

}

@media (min-width:45rem) {

}

<div />
`;

exports[`Modal renders correctly 1`] = `
.c0 {
  all: unset;
  display: grid;
  padding: 0.25rem;
  position: absolute;
  top: 1rem;
  right: 1rem;
  background-color: #e0e0e0;
  color: #828282;
  border-radius: 0.25rem;
}

.c0:focus {
  box-shadow: 0 0 0 0.1875rem rgba(66,153,225,0.5);
}

.c0:hover,
.c0:active {
  color: #333333;
}

.c0:active {
  background-color: #bdbdbd;
}

.c1 {
  overflow: auto;
  max-height: calc(100vh - 2rem);
}

.c1 > * {
  padding: 1rem;
  padding-bottom: 0;
}

.c1 > *::after {
  content: "";
  display: block;
  padding-bottom: calc(1rem + env(safe-area-inset-bottom));
}

@media (min-width:45rem) {

}

@media (min-width:45rem) {

}

@media (min-width:45rem) {
  .c1 > *::after {
    padding-bottom: 1rem;
  }
}

<div>
  <button
    className="c0"
    onClick={[Function]}
    title="Close"
  >
    <svg
      fill="none"
      height={20}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width={20}
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="18"
        x2="6"
        y1="6"
        y2="18"
      />
      <line
        x1="6"
        x2="18"
        y1="6"
        y2="18"
      />
    </svg>
  </button>
  <div
    className="c1"
  />
</div>
`;

exports[`Modal renders the modal content if the modal is opened 1`] = `
.c0 {
  all: unset;
  display: grid;
  padding: 0.25rem;
  position: absolute;
  top: 1rem;
  right: 1rem;
  background-color: #e0e0e0;
  color: #828282;
  border-radius: 0.25rem;
}

.c0:focus {
  box-shadow: 0 0 0 0.1875rem rgba(66,153,225,0.5);
}

.c0:hover,
.c0:active {
  color: #333333;
}

.c0:active {
  background-color: #bdbdbd;
}

.c1 {
  overflow: auto;
  max-height: calc(100vh - 2rem);
}

.c1 > * {
  padding: 1rem;
  padding-bottom: 0;
}

.c1 > *::after {
  content: "";
  display: block;
  padding-bottom: calc(1rem + env(safe-area-inset-bottom));
}

@media (min-width:45rem) {

}

@media (min-width:45rem) {

}

@media (min-width:45rem) {
  .c1 > *::after {
    padding-bottom: 1rem;
  }
}

<div>
  <button
    className="c0"
    onClick={[Function]}
    title="Close"
  >
    <svg
      fill="none"
      height={20}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width={20}
      xmlns="http://www.w3.org/2000/svg"
    >
      <line
        x1="18"
        x2="6"
        y1="6"
        y2="18"
      />
      <line
        x1="6"
        x2="18"
        y1="6"
        y2="18"
      />
    </svg>
  </button>
  <div
    className="c1"
  >
    <section>
      This is the modal content.
    </section>
  </div>
</div>
`;
